python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -216,22 +213,39 @@ def _format_in_threadpool(
             as batches.
         collate_fn: A function to apply to each data batch before returning it.
         num_threadpool_workers: The number of threads to use in the threadpool.
+        metrics_tag: Tag for emitting iteration metrics to _StatsActor.
     """
 
+    last_stats_update = [0]  # hack to make this variable mutable
+    stats_update_lock = threading.Lock()
+
     def threadpool_computations_format_collate(
         batch_iter: Iterator[Batch],
     ) -> Iterator[Batch]:
         # Step 4a: Format the batches.
         formatted_batch_iter = format_batches(
-            batch_iter, batch_format=batch_format, stats=stats
+            batch_iter,
+            batch_format=batch_format,
         )
 
         # Step 4b: Apply the collate function if applicable.
         if collate_fn is not None:
             formatted_batch_iter = collate(
                 formatted_batch_iter, collate_fn=collate_fn, stats=stats
             )
-        yield from formatted_batch_iter
+
+        for batch in formatted_batch_iter:
+            yield batch
+            # Update stats in here to avoid blocking main
+            # iteration thread with task submission overhead.
+            if stats_update_lock.acquire(blocking=False):
+                if (
+                    time.time() - last_stats_update[0]
+                    >= STATS_ACTOR_UPDATE_INTERVAL_SECONDS
+                ):
+                    update_stats_actor_iter_metrics(stats, metrics_tag)
+                    last_stats_update[0] = time.time()
+                stats_update_lock.release()
```

Comment:
I ended up adding this here since `format_batches` is also used in other places, please let me know if there's anything wrong with this

Reply:
it's kinda weird to couple stats updating code with batch formatting code. Especially the thread pool and lock would make the code hard to read. I'd prefer keeping it in the main thread given the 1% overhead, or using a dedicated component (e.g., StatsManager) with a dedicated thread to update the stats periodically.

Reply:
I asked for this:

> Hmm the 1% overhead is not too bad, but I think it's probably best to put it in a background thread still (to allow shorter interval and make the code a bit more robust).

The robustness benefit is worth the code readability to me but happy to sync offline about it.